File: examples/node-express/msMiddleware.js
import fetch from 'node-fetch'

function scrubHeaders(headers, keys = ['authorization', 'cookie']) {
  const out = {}
  for (const [k, v] of Object.entries(headers || {})) {
    out[k] = keys.includes(k.toLowerCase()) ? '***' : v
  }
  return out
}

export function createMiniSentryErrorMiddleware({ baseUrl = 'http://localhost:8000', token, release = '1.0.0', environment = 'development', app = 'node-express-app', scrubHeaderKeys = ['authorization', 'cookie'] } = {}) {
  if (!token) throw new Error('MiniSentry: token is required')
  return function miniSentryErrorMiddleware(err, req, res, next) {
    try {
      const payload = {
        level: 'error',
        message: err?.message || String(err),
        stack: err?.stack,
        release,
        environment,
        app,
        extra: {
          method: req.method,
          path: req.originalUrl || req.url,
          query: req.query,
          headers: scrubHeaders(req.headers, scrubHeaderKeys),
          ip: req.ip,
        },
      }
      // fire and forget
      void fetch(`${baseUrl}/api/events/ingest/token/${token}/`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        keepalive: true,
      }).catch(() => {})
    } catch {}
    next(err)
  }
}

File: examples/ruby-rails/ms_middleware.rb
# frozen_string_literal: true

# Rack middleware for Rails/Rack apps to forward unhandled errors to Mini Sentry
# Usage (Rails):
#   # config/application.rb or an initializer
#   require Rails.root.join('examples', 'ruby-rails', 'ms_middleware')
#   config.middleware.insert_after ActionDispatch::ShowExceptions, MiniSentryRack

require 'net/http'
require 'json'

class MiniSentryRack
  def initialize(app)
    @app = app
    @base = ENV['MS_BASE'] || 'http://localhost:8000'
    @token = ENV['MS_TOKEN'] || 'PASTE_INGEST_TOKEN'
  end

  def call(env)
    @app.call(env)
  rescue => e
    begin
      uri = URI("#{@base}/api/events/ingest/token/#{@token}/")
      payload = {
        message: e.message,
        level: 'error',
        stack: e.full_message(highlight: false, order: :bottom),
        release: ENV['MS_RELEASE'] || '1.0.0',
        environment: ENV['RAILS_ENV'] || ENV['RACK_ENV'] || 'development',
        app: 'rails-example',
      }
      Net::HTTP.start(uri.host, uri.port) do |http|
        req = Net::HTTP::Post.new(uri, 'Content-Type' => 'application/json')
        req.body = JSON.dump(payload)
        http.request(req)
      end
    rescue
      # swallow reporting failures
    end
    raise
  end
end

